druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,240 @@ use crate::kurbo::Point;
 
 use crate::keyboard::KeyModifiers;
 
-/// The state of the mouse for a click, mouse-up, or move event.
+/// Information about the mouse move event.
 #[derive(Debug, Clone, PartialEq)]
-pub struct MouseEvent {
+pub struct MoveEvent {
     /// The location of the mouse in the current window.
     ///
-    /// This is in px units, that is, adjusted for hi-dpi.
+    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
     pub pos: Point,
-    /// Keyboard modifiers at the time of the mouse event.
+    /// Mouse buttons being held down at the time of the event.
+    pub buttons: MouseButtons,
+    /// Keyboard modifiers at the time of the event.
+    pub mods: KeyModifiers,
+}
+
+/// Information about the mouse click event.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ClickEvent {
+    /// The location of the mouse in the current window.
+    ///
+    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
+    pub pos: Point,
+    /// Mouse buttons being held down after the event.
+    /// Thus it will contain the `button` that triggered a mouse-down event,
+    /// and it will not contain the `button` that triggered a mouse-up event.
+    pub buttons: MouseButtons,
+    /// Keyboard modifiers at the time of the event.
     pub mods: KeyModifiers,
     /// The number of mouse clicks associated with this event. This will always
     /// be `0` for a mouse-up event.
-    pub count: u32,
-    /// The currently pressed button in the case of a move or click event,
-    /// or the released button in the case of a mouse-up event.
+    pub count: u8,
+    /// The button that was pressed down in the case of mouse-down,
+    /// or the button that was released in the case of mouse-up.
     pub button: MouseButton,
```

Comment:
okay, thinking about the API:

I think that separating out `MouseButtons` from `MouseButton` solves most of our issue here, and shows a viable way to have a single type cover these three events without any clear inconsistency. There will be two fields, `button` and `count`, that are not used in `MouseMove`, and I think that's fine.

`count` is actually a funny one; the docstring says "will always be 0 on mouse-up", but I'm not sure I actually like that as an implementation; this is something we're going to have to revisit at some point, in that some platforms might provide their own double/triple click counting, and on some platforms we might need to do it ourselves. We probably do want to trust the platform, though, so that we are respecting whatever settings the user has on that platform.

My gut tells me that mouse-up should include count, though? (but this is unrelated to current PR).

Reply:
`MouseMove` would also have a useless `focused` that will be introduced in #842, however even so I'm also thinking a single event is probably the way to go.

As for `count`, I have additional thoughts on it that I describe in #859. The short of it is that yes I think `count` should be there for `MouseUp` but it's not there right now on any platfrom besides macOS so I just wrote the docs to reflect that reality and made macOS compatible with that. This would be a short-lived change in light of #859.